packages/plugin-ext/doc/how-to-add-new-custom-plugin-api.md

```diff
@@ -0,0 +1,281 @@
+            defaultApi = apiFactory(emptyPlugin);
+        }
+
+        return defaultApi;
+    };
+}
+
+function findPlugin(filePath: string): Plugin | undefined {
+    return plugins.getAllPlugins().find(plugin => filePath.startsWith(plugin.pluginFolder));
+}
+```
+
+## Implement your API object
+
+We create a dedicated API object for each individual plugin as part of the module loading process.
+Each API object is returned as part of the module loading process if a script imports `@bar/foo` and should therefore match the API definition that we provided in the `*.d.ts` file.
+Multiple imports will not lead to the creation of multiple API objects as we cache it in our custom `overrideInternalLoad` function.
+
+Example `foo-init.ts` (continued):
+
+```typescript
+export function createAPIFactory(rpc: RPCProtocol): ApiFactory {
+    const fooExtImpl = new FooExtImpl(rpc);
+    return function (plugin: Plugin): typeof fooBarAPI {
+        const FooBar: typeof fooBarAPI.fooBar = {
+            getFoo(): Promise<fooBarAPI.Foo> {
+                return fooExtImpl.getFooImpl();
+            }
+        }
+        return <typeof fooBarAPI>{
+            fooBar : FooBar
+        };
+    }
+}
+```
+
+In the example above the API object creates a local object that will fulfill the API contract.
+The implementation details are hidden by the object and it could be a local implementation that only lives inside the plugin host but it could also be an implementation that uses the `RPCProtocol` to communicate with the main application to trigger changes, register functionality or retrieve information.
+
+### Implement Main-Ext communication
+
+In this document, we will only highlight the individual parts needed to establish the communication between the main application and the external plugin host.
+For a more elaborate example of an API that communicates with the main application, please have a look at the definition of the [Theia Plugin API](https://github.com/eclipse-theia/theia/blob/master/doc/Plugin-API.md).
+
+First, we need to establish the communication on the RPC protocol by providing an implementation for our own side and generating a proxy for the opposite side.
+Proxies are identified using dedicated identifiers so we set them up first, together with the expected interfaces.
+`Ext` and `Main` interfaces contain the functions called over RCP and must start with `$`.
+Due to the asynchronous nature of the communication over RPC, the result should always be a `Promise` or `PromiseLike`.
+
+Example `common/foo-api-rpc.ts`:
+
+```typescript
+export interface FooMain {
+    $getFooImpl(): Promise<Foo>;
+}
+
+export interface FooExt {
+    // placeholder for callbacks for the main application to the extension
+}
+
+// Plugin host will obtain a proxy using these IDs, main application will register an implementation for it.
+export const PLUGIN_RPC_CONTEXT = {
```

Comment:
Maybe we should use different names than `PLUGIN_RPC_CONTEXT` and `MAIN_RPC_CONTEXT`, because those are exported from `@theia/plugin-ext'` as well. Users might pick the wrong import then.

Reply:
I added a `FOO_` prefix to make it even more obvious, thank you!

Reply:
Ah, the age-old dilemma of when to scope-qualify symbols not because they're needed by the code but to make it easier on humans navigating the code. Some would argue the longer names clutter up the source (FOO is short enough to make that a weak argument in this case). Ultimately, it's very subjective, and as with most such discussions, I try to stay far awau...except of course to highlight the dilemma itself. That part is fun and non-judgmental 😆